### DISCUSSION TITLE
java.util.Optional에 대해서 설명해주세요. 그리고, 사용시 주의사항이 있을까요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#47] [Java]
.

**Comment by J-I-H-O** [comment]
Optional 클래스는 null이 올 수 있는 값을 감싸는 Wrapper 클래스입니다. null을 한 번 감싸주기 때문에 참조했을 때 NPE를 방지할 수 있도록 도와줍니다

- 단, 값의 존재 여부를 판단하지 않고 접근하려 시도하면 NPE는 피할 수 있지만, 대신NoSuchElementException이 발생합니다.
- 또, optional 인스턴스 자체가 null일 수 있기 때문에 코드가 불필요하게 복잡해질 수 있습니다.
```
    @Test
    public void testIfPresent_whenOptionalIsNull(){
        //given
        Optional<String> optional = null;
        String actual = "";
        //when
        if(optional != null && optional.isPresent()){
            actual = optional.get();
        }else{
            actual = "default";
        }
        //then
        assertThat(actual).isEqualTo("default");
    }
```
- 또 Optional을 남용하게 되면 시간적, 공간적 비용이 증가합니다. Optional은 객체를 감싸는 래퍼 클래스이므로 Optional 객체 자체를 저장하기 위한 메모리가 추가로 필요하고, Optional 안에 있는 객체를 얻기 위해서는 Optional 객체를 통해 접근해야 하므로 접근 비용이 증가합니다.

**Comment by 3Juhwan** [comment]
Optional은 Java 8에서 도입된 클래스로, 값이 존재할 수도 있고 없을 수도 있는 상황을 다루는 데 사용됩니다. 이 클래스는 주로 메서드의 반환 타입으로 사용되어, 메서드가 값을 반환하지 못하는 경우에 명시적으로 그 상황을 나타냅니다. null 대신에 Optional을 사용함으로써 NPE를 회피할 수 있습니다. 

주의할 점은 다음과 같습니다. 

- Optional을 사용하면 코드가 복잡해져 가독성이 저하될 수 있습니다. 
- Optional로 값을 감싸고 푸는 과정에서 오버헤드가 발생하고 성능 저하로 이어질 수 있으므로, 필요한 경우에만 사용해야 합니다.

**Comment by le2sky** [comment]
https://yeonyeon.tistory.com/224
https://le2ksy.tistory.com/65
https://dev-coco.tistory.com/178

**Comment by GIVEN53** [comment]
Java 8
객체 wrapping
NPE 회피
코드 복잡성, 가독성 저하, 오버헤드
